File: src/layout.cpp
#include <campaign/layout.hpp>
#include <stdexcept>

using namespace campaign;

using FlagInfo = Descriptor::FlagInfo;
using ByteInfo = Descriptor::ByteInfo;
using DynamicInfo = Descriptor::DynamicInfo;

Layout::Layout()
{
    cursor_ = 0;
    flagCursor_ = 0x1;
}

Layout::Layout(size_t entryCount)
    : Layout()
{
    descriptorMap_.reserve(entryCount);
}

const Descriptor &Layout::operator[](const std::string &id) const
{
    return descriptorMap_.at(id);
}

void Layout::pushFlag(const std::string &id)
{
    descriptorMap_.insert({id, Descriptor(FlagInfo(cursor_, flagCursor_))});

    stepFlag();
}

void Layout::pushFlag()
{
    stepFlag();
}

void Layout::pushByte(const std::string &id)
{
    stepToFreeByte();

    descriptorMap_.insert({id, Descriptor(ByteInfo(cursor_))});

    stepByte();
}

void Layout::pushByte()
{
    stepToFreeByte();

    stepByte();
}

void Layout::pushDynamic(const std::string &id, size_t size)
{
    stepToFreeByte();

    descriptorMap_.insert({id, Descriptor(DynamicInfo(cursor_, size))});

    stepBlock(size);
}

void Layout::pushDynamic(size_t size)
{
    stepToFreeByte();

    stepBlock(size);
}

bool Layout::entryExists(const std::string id) const
{
    return descriptorMap_.contains(id);
}

size_t Layout::getEntryCount() const
{
    return descriptorMap_.size();
}

size_t Layout::getDataSize() const
{
    if (flagCursor_ == 0x1)
    {
        return cursor_;
    }

    return cursor_ + 1;
}

Layout::EntryInfo Layout::getEntryInfo(size_t index) const
{
    size_t size = descriptorMap_.size();

    if (index >= size)
    {
        throw std::out_of_range("Trying to get layout entry at " + std::to_string(index) + " when there's only " + std::to_string(size) + ".");
    }

    auto iterator = std::next(descriptorMap_.begin(), index);

    return Layout::EntryInfo(iterator->first, iterator->second);
}

const Descriptor &Layout::getDescriptor(const std::string &id) const
{
    return descriptorMap_.at(id);
}

Descriptor::FlagInfo Layout::getFlagInfo(const std::string &id) const
{
    return descriptorMap_.at(id).getFlagInfo();
}

Descriptor::ByteInfo Layout::getByteInfo(const std::string &id) const
{
    return descriptorMap_.at(id).getByteInfo();
}

Descriptor::DynamicInfo Layout::getDynamicInfo(const std::string &id) const
{
    return descriptorMap_.at(id).getDynamicInfo();
}

void Layout::stepFlag()
{
    if (flagCursor_ == (0x1 << 7))
    {
        cursor_++;
        flagCursor_ = 0x1;
        return;
    }

    flagCursor_ <<= 1;
}

void Layout::stepByte()
{
    flagCursor_ = 0x1;
    cursor_++;
}

void Layout::stepBlock(size_t count)
{
    flagCursor_ = 0x1;
    cursor_ += count;
}

void Layout::stepToFreeByte()
{
    if (flagCursor_ == 0x1)
    {
        return;
    }

    stepByte();
}

File: bindings/csharp/Campaign/Layout.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using Campaign.API;

namespace Campaign
{
    public class Layout : IDisposable, IEnumerable<Layout.EntryInfo>
    {
        private IntPtr _layoutPtr;


        public Layout(int descriptorCount)
        {
            LayoutAPI.Create(descriptorCount, out _layoutPtr).ValidateAPICall();
        }

        internal Layout(IntPtr layoutPtr)
        {
            _layoutPtr = layoutPtr;
        }

        ~Layout()
        {
            DisposeInternal();
        }


        void IDisposable.Dispose()
        {
            DisposeInternal();
            GC.SuppressFinalize(this);
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }


        internal IntPtr ptr => _layoutPtr;


        public void PushFlag(string id)
        {
            LayoutAPI.PushFlag(_layoutPtr, id).ValidateAPICall();
        }

        public void PushFlag()
        {
            LayoutAPI.PushFlag(_layoutPtr).ValidateAPICall();
        }

        public void PushByte(string id)
        {
            LayoutAPI.PushByte(_layoutPtr, id).ValidateAPICall();
        }

        public void PushByte()
        {
            LayoutAPI.PushByte(_layoutPtr).ValidateAPICall();
        }

        public void PushDynamic(string id, int size)
        {
            LayoutAPI.PushDynamic(_layoutPtr, id, size).ValidateAPICall();
        }

        public void PushDynamic(int size)
        {
            LayoutAPI.PushDynamic(_layoutPtr, size).ValidateAPICall();
        }

        public bool EntryExist(string id)
        {
            LayoutAPI.EntryExists(_layoutPtr, id, out bool exists).ValidateAPICall();

            return exists;
        }

        public bool EntryExist(string id, Descriptor.InfoType infoType)
        {
            LayoutAPI.EntryExists(_layoutPtr, id, infoType, out bool exists).ValidateAPICall();

            return exists;
        }

        public int GetEntryCount()
        {
            LayoutAPI.GetEntryCount(_layoutPtr, out var entryCount).ValidateAPICall();

            return entryCount;
        }

        public int GetDataSize()
        {
            LayoutAPI.GetDataSize(_layoutPtr, out var dataSize).ValidateAPICall();

            return dataSize;
        }

        public EntryInfo GetEntryInfo(int index)
        {
            LayoutAPI.GetEntryInfo(_layoutPtr, index, out var nativeEntry).ValidateAPICall();

            try
            {
                return nativeEntry.TransferToManaged();
            }
            finally
            {
                nativeEntry.Dispose();
            }
        }

        public Descriptor GetDescriptor(string id)
        {
            LayoutAPI.GetDescriptor(_layoutPtr, id, out var descriptor).ValidateAPICall();

            return descriptor;
        }

        public Descriptor.FlagInfo GetFlagInfo(string id)
        {
            LayoutAPI.GetFlagInfo(_layoutPtr, id, out var flagInfo).ValidateAPICall();

            return flagInfo;
        }

        public Descriptor.ByteInfo GetByteInfo(string id)
        {
            LayoutAPI.GetByteInfo(_layoutPtr, id, out var byteInfo).ValidateAPICall();

            return byteInfo;
        }

        public Descriptor.DynamicInfo GetDynamicInfo(string id)
        {
            LayoutAPI.GetDynamicInfo(_layoutPtr, id, out var dynamicInfo).ValidateAPICall();

            return dynamicInfo;
        }

        public IEnumerator<EntryInfo> GetEnumerator()
        {
            var count = GetEntryCount();
            for (int i = 0; i < count; i++)
            {
                yield return GetEntryInfo(i);
            }
        }

        private void DisposeInternal()
        {
            if (_layoutPtr == IntPtr.Zero)
            {
                return;
            }

            LayoutAPI.Delete(_layoutPtr).ValidateAPICall();

            _layoutPtr = IntPtr.Zero;
        }


        [StructLayout(LayoutKind.Sequential)]
        public readonly struct EntryInfo
        {
            public readonly string id;

            public readonly Descriptor descriptor;


            public EntryInfo(in string id, in Descriptor descriptor)
            {
                this.id = id;
                this.descriptor = descriptor;
            }
        }
    }
}